feat(ejs): render and attach shadowroot in the same tick.

quarkus opened the pull request:
Currently we have an animation frame between connection of an element and the first render cycle.

I ran into a bug today while rendering a Shadow Element as we actually attach the ShadowRoot on connected but actually render not until the requested update is performed.

That leaves the element in a state where the light dome is kind of "detached" from the DOM.

That leads to eg. nested elements having NO offsetParent and no dimensions at all:
<img width="880" height="985" alt="screenshot_mh 2025-09-04 um 15 24 33" src="https://github.com/user-attachments/assets/2c177b98-3e32-41ca-9d82-4323d539dae5" />


With this change i would like to keep the SR Attache and the first render in the same tick.

Is there any reason why we should not do it like this @eddyloewen !?

quarkus committed:
feat(ejs): render and attach shadowroot in the same tick.

KinG-DE commented on src/TemplateElement.js:
```diff
@@ -48,6 +47,12 @@ class TemplateElement extends StyledElement {
     update(options) {
+        let firstRender = this._options.shadowRender && !this.shadowRoot;
+        if (firstRender) {
+            this.attachShadow({ mode: 'open' });
```
Is there a need to  verify if shadow root creation actually succeeded?

eddyloewen replied:
@KinG-DE !this.shadowRoot should do exactly that

eddyloewen commented on src/TemplateElement.js:
```diff
@@ -48,6 +47,12 @@ class TemplateElement extends StyledElement {
      * @param {PropertyUpdateOptions} options
      */
     update(options) {
+        let firstRender = this._options.shadowRender && !this.shadowRoot;
```
@quarkus this check will now run for every update cycle. There can be A LOT of update cycles... Is this ok performance wise?

quarkus replied:
i guess we should set a simble "attached"  flag somewhere, but yes, guess it makes sense to keep it as simple as possible.

quarkus replied:
Hey @eddyloewen ! I challenged the idea but AI suggests that there is no performance to gain. 

both boolean and the falsy check will be rather fast:
Accessing this.shadowRoot is a direct property lookup on the object, which is one of the fastest operations in JavaScript. It is a fundamental part of the object model and is highly optimized by browser engines.

It also pointed out that adding more flags will probably introduce new bugs which is probably also true :D

quarkus replied:
ps.: i also moved it to the render function. maybe it makes more sense in combination with the adoptStyleSheets call.

eddyloewen commented on src/TemplateElement.js:
```diff
@@ -48,6 +47,12 @@ class TemplateElement extends StyledElement {
+        if (firstRender) {
+            this.attachShadow({ mode: 'open' });
+            // adopt stylesheets
+            this.adoptStyleSheets();
```
@quarkus where does this come from? This call was not here before...

quarkus replied:
That actually comes from the "StyledElements" connectedCallback which we extend from here.

Since i moved the "attach" call here i have to somehow "retrigger" the adoption after the shadowroot is attached. Otherwise the  styles are adopted into nohwere ...

quarkus replied:
Please check again @eddyloewen . I rearranged the code and also removed it from the StyledElements connected as it actually needs a root to make sense.

quarkus committed:
feat(ejs): move attach check to the actual renderTemplate function

quarkus committed:
feat(ejs): moved external style initialization to a dedicated function (initStyleAdoption) which makes things hopefully better to understand.

also removed the code from StyledElements connected as it needs a shadowRoot anyways to actually make sense.
